### ISSUE TITLE
Extend type checking to use the injectivity information

### ISSUE THREAD

**Issue by jstolarek** [jstolarek/ghc#69]
There was some discussion [on trac](https://ghc.haskell.org/trac/ghc/ticket/6018#comment:89) and Simon offered to do this.

**Comment by jstolarek** [comment]
Simon's patch is ready:

https://gist.github.com/jstolarek/95a6b07de2ab95656bc3

**Comment by jstolarek** [comment]
There are several issues with implementation for closed type families:
1. Overloading causes some issues:

``` haskell
type family Bak a = r | r -> a where
     Bak Int  = Char
     Bak Char = Int
     Bak a    = a

bak :: Bak a -> Bak a
bak x = x

bakapp2 :: Double
bakapp2 = bak 1.0
```

produces:

```
T6018.hs:110:11:
    Couldn't match expected type ‘Double’ with actual type ‘Bak a1’
    The type variable ‘a1’ is ambiguous
    In the expression: bak 1.0
    In an equation for ‘bakapp2’: bakapp2 = bak 1.0

T6018.hs:110:15:
    Occurs check: cannot construct the infinite type: a1 ~ Bak a1
    The type variable ‘a1’ is ambiguous
    Expected type: Rational -> Bak a1
      Actual type: Rational -> a1
    In the first argument of ‘bak’, namely ‘1.0’
    In the expression: bak 1.0
    In an equation for ‘bakapp2’: bakapp2 = bak 1.0
```

Putting a type signature on `1.0` makes the problem go away. Another instance of the same issue:

``` haskell
type family Bar a = r | r -> a where
    Bar Int  = Bool
    Bar Bool = Int
    Bar Bool = Char

bar :: Bar a -> Bar a
bar x = x

barapp2 :: Int
barapp2 = bar 1
```

produces:

```
T6018.hs:125:11:
    Couldn't match expected type ‘Int’ with actual type ‘Bar a0’
    The type variable ‘a0’ is ambiguous
    In the expression: bar 1
    In an equation for ‘barapp2’: barapp2 = bar 1
```

2.

``` haskell
type family Bar a = r | r -> a where
    Bar Int  = Bool
    Bar Bool = Int
    Bar Bool = Char

bar :: Bar a -> Bar a
bar x = x

barapp2 :: Char
barapp2 = bar 'c'
```

produces

```
T6018failclosed10.hs:17:11:
    Couldn't match type ‘Int’ with ‘Char’
    Expected type: Char
      Actual type: Bar Bool
    In the expression: bar 'c'
    In an equation for ‘barapp2’: barapp2 = bar 'c'

T6018failclosed10.hs:17:15:
    Couldn't match type ‘Int’ with ‘Char’
    Expected type: Bar Bool
      Actual type: Char
    In the first argument of ‘bar’, namely ‘'c'’
    In the expression: bar 'c'
    In an equation for ‘barapp2’: barapp2 = bar 'c'
```

These two errors taken together are confusing, though probably both are correct on their own.

**Comment by jstolarek** [comment]
Second problem can be solved by dumping unreachable equations - see #9085

**Comment by jstolarek** [comment]
These two problems now have their own separate issue numbers: #111 and #112
